src/modules/mc_pos_control/MulticopterPositionControl.cpp

```diff
@@ -317,7 +317,15 @@ void MulticopterPositionControl::Run()
 
 		if (_vehicle_control_mode.flag_multicopter_position_control_enabled) {
 
-			const bool is_trajectory_setpoint_updated = _trajectory_setpoint_sub.update(&_setpoint);
+			bool is_trajectory_setpoint_updated = false;
+
+			if (_vehicle_control_mode.flag_control_offboard_enabled) {
+				is_trajectory_setpoint_updated = _offboard_trajectory_setpoint_sub.update(&_setpoint);
```

Comment:
Share the full log? I'll take a closer look.

Reply:
Sorry I can't share right now, but maybe later. I did notice the topic `trajectory_setpoints` was completely missing although I armed in position mode then switched to offboard. This looks like a `flight_manager` failure to me since it's the only publisher of `trajectory_setpoints` in this case.

Reply:
After looking closer at the logs it appears most offboard_trajectory_setpoints sent were never receieved.
I suspect problems with the FTDI adapter not keeping up and causing timesync and dropout issues, regardless px4 should be able to handle it.
Here's the steps taken in this log: https://logs.px4.io/plot_app?log=9d7d039f-6c26-4a9c-8d13-a02aced4e588
1. arm in stablized (observe no trajectory_setpoints are ever logged)
2. publish `offboard_control_mode` and `offboard_traj_setpoints` (only `offboard_control_mode` is received)
3. switch to offboard mode (observe `vehicle_local_position.z` climbs although no setpoints are given)
4. eventually quad stops moving and holds its position
5. `offboard_trajectory_setpoints` begin to be received and suddenly vehicle_local_position_setpoint.acceleration[2] = 100
	This happens at line 407 https://github.com/PX4/PX4-Autopilot/blob/3e7a58d04c3bdd070b63bfe4caeb8c909475931e/src/modules/mc_pos_control/MulticopterPositionControl.cpp#L407
This is possible because the takeoff status is `ready_for_takeoff` and never proceeds to `flying`
Here's a plot showing what's described
![image](https://user-images.githubusercontent.com/9113272/141006576-2a10d662-fa3c-4901-9452-a6e2514f828d.png)

Here's another log where the same behavior happened.
https://logs.px4.io/plot_app?log=58a7400d-6e44-46b8-8dcc-d7add2d9029d

These scenarios all happened at the end of some other successful flight testing with the same px4 version.
Here's one of several successful offboard tests where `invalid_setpoints` setpoints didn't get logged and `offboard_trajectory_setpoints` were consistently received. 
https://logs.px4.io/plot_app?log=56b11c56-5055-4583-84a3-eda7b6470f19
One observation is it had one `trajectory_setpoint` at the start

I'm trying to replicate in sim currently, but I can't figure out how in step 3, quad climbs in altitude and `trajectory_setpoints` aren't published.
I have 2 possible ideas:
1. an old `_setpoint` value is being used and doesn't update
2. local_position drifts downwards causing quad to rise due to using `ekf2_hgt_mode=GPS`

Reply:
I found the bug and finally was able to replicate in sim after logging from boot outside to confirm what I thought was going on, it's my 2nd possible idea above. The problem is the mc_pos_controller goes into failsafe and since the local_position drifts down, it causes the upwards velocity to set to the difference between 0 and the drifted `_pos` value. This happens here: https://github.com/PX4/PX4-Autopilot/blob/3e7a58d04c3bdd070b63bfe4caeb8c909475931e/src/modules/mc_pos_control/PositionControl/PositionControl.cpp#L115  Then the thrust gets applied to 100% and the quad rises very quickly. Then after the failsafe hystersis catches up, the setpoint gets set to nan and the controller can no longer cause accelerations so the quad holds it position. In this state, the quad is holding it's position, but the takeoff state is still `ready_for_takeoff`. Once `offboard_trajectory_setpoints` begin to be received for the first time this code is applied and the quad falls: https://github.com/PX4/PX4-Autopilot/blob/3e7a58d04c3bdd070b63bfe4caeb8c909475931e/src/modules/mc_pos_control/MulticopterPositionControl.cpp#L407 I was able to replicate this in sim, but I had to modify mc_pos_controller to get it to work. I changed the update on `position_control_enabled` to `control_offboard_enabled` and added a delay to the hysteresis so the sim would allow thrust to be applied longer since the physical acceleration in sim is slower than my quad. Also I needed a custom simulator to initialize px4, then drop down so the `local_pos` would begin below the ground.

Reply:
If anyone would like to replicate this problem, I made an example repo here: https://github.com/benjinne/px4_offtest
I made some modifications to demonstrate the problem:
1. Added a node to px4_ros_com to only publish `offboard_control_mode`
   -https://github.com/PX4/px4_ros_com/commit/72b9efdce3dc9d855f5c2b7e5d357acd2f50d895
2. Modified PX4 to only run the position controller when offboard is enabled
3. Added delay to position controller's failsafe hystersis to allow time for quad to rise in sim since it rises in the real testing, but not in sim. Either way, 100% thrust get applied briefly which is a bug.
   -https://github.com/PX4/PX4-Autopilot/commit/447e0eb091816043a04a1680969f38ce1ff82cef 

To replicate the bug, clone my repo and follow these steps:

Terminal 1:
colcon build --symlink-install
cd src/px4
make px4_sitl_rtps none_iris

Terminal 2:
source install/setup.bash
micrortps_agent -t UDP

Terminal 3:
cd src/px4
DONT_RUN=1 make px4_sitl_rtps gazebo
source Tools/setup_gazebo.bash $(pwd) $(pwd)/build/px4_sitl_rtps
gazebo Tools/sitl_gazebo/worlds/empty.world --verbose

Terminal 4:
cd src/px4
source Tools/setup_gazebo.bash $(pwd) $(pwd)/build/px4_sitl_rtps
gz model --verbose --spawn-file=./Tools/sitl_gazebo/models/iris/iris.sdf --model-name=iris -x 0.0 -y 0.0 -z 50.0

Terminal 5:
source install/setup.bash
ros2 run px4_ros_com offboard_only_control_mode

QGC:
Enable virtual joysticks
Switch to Stabalized mode
Arm and Offboard
Observe logs show throttle spike to 100%. This can be made longer by extending the length of hystersis

This bug is not a result of this PR, although it can be addressed here. My one possible solution is to remove the `reset_setpoint_to_nan` in the initialization of mc_pos_controller. This prevents the failsafe routine from being called since `control_update()` won't return false right away and `invalid_setpoints` doesn't get printed. @dagar Are you okay with removing this line: https://github.com/PX4/PX4-Autopilot/blob/7653bd17575d88545969e5d127c74fa7ca86498f/src/modules/mc_pos_control/MulticopterPositionControl.cpp#L56

Reply:
Also, I've done several flight tests in position mode and offboard mode outside and in the sim with this line removed: https://github.com/PX4/PX4-Autopilot/blob/7653bd17575d88545969e5d127c74fa7ca86498f/src/modules/mc_pos_control/MulticopterPositionControl.cpp#L56 I have not identified any problems with this solution other than it could be easily re-added and cause the problem again